iup/eval_objects.py

```diff
@@ -0,0 +1,268 @@
+    def get_metric(self, metric_type):
+        """
+        Calculate metric based on `metric_type` of
+        joined dataframe from data and prediction.
+        """
+        if metric_type == 'mspe':
+            return self.get_mspe()
+        elif metric_type == 'mean_bias':
+            return self.get_mean_bias()
+        elif metric_type == 'eos_abe':
+            return self.get_eos_abe()
+        else:
+            raise Exception(f'Does not support {metric_type}')
+
+    # metric can be directly called too
+    def get_mspe(self):
+        """
+        Calculate MSPE from joined data
+        ----------------------
+        Input: self (joined data)
+        Return: pl.DataFrame with MSPE and the forecast start date
+
+        """
+        return self.with_columns(
+            spe = (pl.col('estimate') - pl.col("estimate_right"))**2
+            ).with_columns(
+                mspe = pl.col('spe').mean(),
+            ).filter(
+                pl.col('date') == pl.col('date').min(),
+            ).rename(
+                {'date':'forecast_start'}
+            ).select(
+                'forecast_start','mspe'
+            )
+
+    def get_mean_bias(self):
+        """
+        Calculate Mean bias from joined data.
+        Note the bias here is not the classical bias calculated from the posterior distribution.
+
+        The bias here is defined as: at time t,
+        bias = -1 if pred_t < data_t; bias = 0 if pred_t == data_t; bias = 1 if pred_t > bias_t
+
+        mean_bias = sum of the bias across time/length of data
+        -------------------------
+        Input: self (joined data)
+        Return: pl.DataFrame with mean bias and the forecast start date
+        """
+        joined = self.with_columns(
+                pl.when(pl.col('estimate') < pl.col('estimate_right')
+                ).then(-1)
+                .when(pl.col('estimate') == pl.col('estimate_right'))
+                .then(0)
+                .otherwise(1).alias('bias')
+            )
+
+        m_bias = pl.DataFrame({'forecast_start':joined['date'].min(),
+                           'mbias':joined['bias'].sum()/joined.shape[0]})
+
+        return m_bias
+
+    def get_eos_abe(self):
+
+        """
+        Calculate the absolute error of the total uptake at the end of season between data and prediction.
+        Maybe can belong to TimelessPointMetric because not every date needs to be matched,
+        but the situation when the last date does not match needs to be defined.
+        -------------------
+        Input: self (joined data)
+        Return: pl.DataFrame with absolute error in the total uptake between data and prediction
+                and the forecast end date.
+        """
+        joined = self.with_columns(
+                cumu_data = pl.col('estimate').cum_sum(),
+                cumu_pred = pl.col('estimate_right').cum_sum()
```

Comment:
Looks like this is assuming that the input PointForecast objects `data` and `pred` are derived from type IncidentUptakeData. If they were derived from CumulativeUptakeData, the `.cum_sum()` would be unnecessary. There should be some way to check or enforce which UptakeData type is used to make a PointForecast object.

Reply:
Certain metrics will only make sense on certain kinds of forecasts

That could either be with checking the classes of the inputs, or by looking at the `target` column of a `Forecast` object, and making sure that it is, say `"incident_uptake"`